massa-deferred-calls/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+        [DEFERRED_CALLS_PREFIX][slot][SLOT_TOTAL_GAS] -> u64 // total gas booked for a slot (optional, default 0, deleted when set to 0)
+        [DEFERRED_CALLS_PREFIX][slot][SLOT_BASE_FEE] -> u64 // deleted when set to 0
+        [DEFERRED_CALLS_PREFIX][slot][CALLS_TAG][id][CALL_FIELD_X_TAG] -> AsyncCall.x // call data
+    */
+
+    // TODO pass args
+    pub fn new(db: ShareableMassaDBController) -> Self {
+        Self {
+            db,
+            call_serializer: DeferredCallSerializer::new(),
+            call_id_serializer: DeferredCallIdSerializer::new(),
+            call_deserializer: DeferredCallDeserializer::new(THREAD_COUNT),
+            call_id_deserializer: DeferredCallIdDeserializer::new(),
+            registry_changes_deserializer: DeferredRegistryChangesDeserializer::new(
+                THREAD_COUNT,
+                MAX_ASYNC_GAS,
+                DEFERRED_CALL_MAX_POOL_CHANGES,
+            ),
+            registry_changes_serializer: DeferredRegistryChangesSerializer::new(),
+        }
+    }
+
+    pub fn get_slot_calls(&self, slot: Slot) -> DeferredSlotCalls {
+        let mut to_return = DeferredSlotCalls::new(slot);
+        let key = deferred_slot_call_prefix_key!(slot.to_bytes_key());
+
+        // cache the call ids to avoid duplicates iteration
+        let mut temp = Vec::new();
+
+        for (serialized_key, _serialized_value) in self.db.read().prefix_iterator_cf(STATE_CF, &key)
+        {
+            if !serialized_key.starts_with(&key) {
+                break;
+            }
+
+            let rest_key = &serialized_key[key.len()..];
+
+            let (_rest, call_id) = self
+                .call_id_deserializer
+                .deserialize::<DeserializeError>(&rest_key)
+                .expect(KEY_DESER_ERROR);
+
+            if temp.contains(&call_id) {
+                continue;
+            }
+
+            temp.push(call_id.clone());
+
+            if let Some(call) = self.get_call(&slot, &call_id) {
+                to_return.slot_calls.insert(call_id, call);
+            }
+        }
+
+        to_return.slot_base_fee = self.get_slot_base_fee(&slot);
+        to_return.slot_gas = self.get_slot_gas(&slot);
+        to_return.total_gas = self.get_total_gas();
+
+        to_return
+    }
+
+    /// Returns the DeferredCall for a given slot and id
+    pub fn get_call(&self, slot: &Slot, id: &DeferredCallId) -> Option<DeferredCall> {
+        let mut buf_id = Vec::new();
+        self.call_id_serializer
+            .serialize(id, &mut buf_id)
+            .expect(DEFERRED_CALL_SER_ERROR);
+        let key = deferred_call_prefix_key!(buf_id, slot.to_bytes_key());
+
+        let mut serialized_call: Vec<u8> = Vec::new();
+        for (serialized_key, serialized_value) in self.db.read().prefix_iterator_cf(STATE_CF, &key)
+        {
+            if !serialized_key.starts_with(&key) {
+                break;
+            }
+
+            serialized_call.extend(serialized_value.iter());
```

Comment:
are we sure that the data is concatenated in the right order to be deserialized afterwards?
Even if we assume "alphabetical order", it creates strong coupling which might be error-prone in the future

Reply:
I think it's ok, `put_entry()` and `deserialize()` are in the same order 
https://github.com/massalabs/massa/blob/f43fdea85d9d84bf543642d37088910f83bf2410/massa-deferred-calls/src/lib.rs

https://github.com/massalabs/massa/blob/f43fdea85d9d84bf543642d37088910f83bf2410/massa-deferred-calls/src/call.rs#L165


and insertion in db and get_call are unit tested in scenario_mandatories